### PR TITLE
Stream commands for Redis 5

### PR DESCRIPTION
**Pull request by jrmhaig**
After raising #141 I started implementing commands for streams (see https://redis.io/topics/streams-intro). I have implemented the commands:
* `xadd`
* `xlen`
* `xrange`
* `xrevrange`

There are more to do, listed [here](https://github.com/jrmhaig/mock_redis/blob/master/lib/mock_redis/stream.rb#L6-L18) but I don't know if I am going to get any further myself. This might be a useful starting point if someone wants to carry on.

### CODE REVIEW

**Comment by jrmhaig** [comment]
The tests fail because they depend on a Redis 5 server.

**Comment by jrmhaig** [comment]
It is now running the tests against Redis 5, which needs to be built as it is still a release candidate. You might want to wait until there is an official package that can be installed by apt.

Also, as I mentioned in #141, a couple of the tests are failing due to the automatically generated ids being one or two milliseconds out between the real and the mocked server. There may not be any way to fix this without making changes to how the tests are run.

**Comment by jrmhaig** [comment]
How do you feel about this? https://github.com/brigade/mock_redis/pull/142/commits/2cc12b85db302a8bbf4588044673e110676e0f5c

If `@redises._gsub(/pattern/, 'substitution')` is called in a test then the string values returned by the mocked and real Redis are compared after `gsub(/pattern/, 'substitution')` is called on both. This is cleared before each test so that they do not affect each other.

Also, the `xadd` and `xlen` test files each have a sleep of up to one second so that if there are any differences of one millisecond between keys then they will only ever change in the final digit.

With these two changes the tests can be made to be reliable. However, it does add sleeps to the tests, which I don't really like.

File: .travis.yml [hunk 1]
```diff
@@ -4,26 +4,34 @@ sudo: false
...
-services:
-  - redis-server
+#services:
+#  - redis-server
 
 rvm:
   - 2.0
   - 2.1
   - 2.2
   - 2.3.1
```
Comment (sds) [inline .travis.yml]: Can get rid of 2.3.1 if you're adding 2.3.7

**Comment by sds** [comment]
I'm not crazy about the solution but I don't have better alternatives to suggest. Happy to move forward with what you've put together once the tests are passing.

Thanks for your work on this!

**Comment by jrmhaig** [comment]
I have resolved conflicts and removed Ruby 2.3.1 from the Travis configuration, as you suggested. The tests are now all passing except for  2.1, which appears to have hung at the start.

Bear in mind that this is now running against Redis 5.0, which is currently unreleased (using release candidate 3) and so it needs to be downloaded and built by Travis each time (see [`start-redis-five.sh`](https://github.com/jrmhaig/mock_redis/blob/master/start-redis-five.sh). This will need to be updated for any new release candidates and the final release so you might decide to wait until it is released and becomes available to the Travis servers via apt before merging this PR.

**Comment by sds** [comment]
Thanks for your hard work here, @jrmhaig. I think we'll take your advice and wait for Redis 5 to be properly released before we merge this. As I'm not personally following that development closely, would appreciate you jumping back in to let us know when it's out.

Thanks!

**Comment by jrmhaig** [comment]
I'm aware that the build failed in Travis after merging from master. It hasn't broken but the wildcard match defined by [`_gsub`](https://github.com/brigade/mock_redis/pull/142/files#diff-735e17b919ec6dbdbe614b8f43ca38c1R16) to get round the problem of the keys being based on millisecond timestamps isn't being used in some cases. I'll have a look at it when I have a moment.

**Comment by sds** [comment]
Looks like the Travis CI build environment now [installs Redis 5](https://docs.travis-ci.com/user/reference/xenial/#databases-and-services) by default.

Appreciate the amount of effort you've put in, @jrmhaig. If you have any bandwidth to take it the rest of the way, would be more than happy to merge!

**Comment by jrmhaig** [comment]
Sorry, I've been a bit distracted but I'll look into it.

Feel free to prod me again if you notice I haven't updated again soon. ;-)

**Comment by sds** [comment]
Thanks @jrmhaig! Happy to merge, however the only option this repo is allowing me to use is Squash and Merge, which would lose all your commits (Rebase and Merge is saying there are conflicts). Did you want to try fixing those conflicts to preserve commit history, or are you fine with me squash merging?

**Comment by jrmhaig** [comment]
OK, I tried a rebase and with hindsight it might have been better to let you squash and merge. Anyway, it is done now. Are you able to merge now?

**Comment by sds** [comment]
For some reason it still doesn't let me:

image

Going to Squash and Merge. Sorry for the trouble!

**Pull request merged**